File: Algorithm-Swift/Algorithm-Swift/BaekJoon/나무자르기.swift
import Foundation

let input = readLine()!.split{$0==" "}.map{Int(String($0))!}
let N = input[0], M = input[1]
let treeArr = readLine()!.split{$0==" "}.map{Int(String($0))!}

var start = 1
var end = treeArr.max()!

while start <= end {
    let mid = (start+end)/2  
    var length = 0
    for i in 0..<N {
      if treeArr[i] > mid {
        length += (treeArr[i] - mid)
      }
    }

    if length >= M {
        start = mid + 1
    } else {
        end = mid - 1
    }
}

print(end)

File: By-Python/baekjoon/[이분탐색]/나무자르기.py
# 나무 자르기
n,m = map(int, input().split())
arr = list(map(int, input().split()))

start, end = 1, max(arr)

while start <= end:
  trims = 0
  mid = (start+end)//2
  for t in arr:
    if t-mid >= 0:
      trims += t-mid
  if trims >= m:
    start = mid + 1
  else:
    end = mid - 1

print(end)